### PR TITLE
feat: improve progress msg in SHOW JOBS for source backfill

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previous discussion here: https://github.com/risingwavelabs/risingwave/pull/18112#discussion_r1741571151

Previously, `progress` in `SHOW JOBS` is a percentile string. After we introduced different type of backfill, things become tricker:

- For some types, like SourceBackfill, we cannot even tell what's the end position, thus no percentile at all.
- We may have different type of backfills in one MV. How to display them together is a problem.
- Snapshot backfill introduced complex string as `progress`: https://github.com/risingwavelabs/risingwave/blob/05b68bcd11a901e2e580f61bdcee799c860b6948/src/meta/src/barrier/creating_job/mod.rs#L130-L175

The idea of this PR is simple: just report separately, and abuse the string.
(Ideas to improve how the `progress` string look like is welcomed)

```rust
match (mv_progress, source_progress) {
    (Some(mv_progress), Some(source_progress)) => {
        format!(
            "MView Backfill: {}, Source Backfill: {}",
            mv_progress, source_progress
        )
    }
    (Some(mv_progress), None) => mv_progress,
    (None, Some(source_progress)) => source_progress,
    (None, None) => "Unknown".to_string(),
}
```

Example:
```
┌────┬──────────────────────────────────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────┐
│ Id │                                      Statement                                       │                            Progress                            │
├────┼──────────────────────────────────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────┤
│ 22 │ CREATE MATERIALIZED VIEW mv AS SELECT * FROM t UNION SELECT v1 FROM s_before_produce │ MView Backfill: 99.99% (3/3), Source Backfill: 0 rows consumed │
└────┴──────────────────────────────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────┘
```

Part of https://github.com/risingwavelabs/risingwave/issues/18338

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
* **#18925** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18925?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/stream/src/executor/source/source_backfill_executor.rs [hunk 1]
```diff
@@ -438,6 +438,8 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
         yield Message::Barrier(barrier);
 
         {
+            // TODO: persist backfilled row count?
+            let mut total_backfilled_rows: u64 = 0;
```
Comment (xxchan) [inline src/stream/src/executor/source/source_backfill_executor.rs:442]: Follow-up here #18983, where I'm not quite satisfied with the solution. 😕

File: src/meta/src/barrier/progress.rs [hunk 2]
```diff
@@ -137,19 +157,52 @@ impl Progress {
...
+        match (mv_progress, source_progress) {
+            (Some(mv_progress), Some(source_progress)) => {
+                format!(
+                    "MView Backfill: {}, Source Backfill: {}",
+                    mv_progress, source_progress
+                )
+            }
+            (Some(mv_progress), None) => mv_progress,
+            (None, Some(source_progress)) => source_progress,
+            (None, None) => "Unknown".to_string(),
```
Comment (xxchan) [inline src/meta/src/barrier/progress.rs:204]: core logic here

**Comment by xxchan** [comment]
> Can we show progress via backfilled offset instead?

Good point. 🤔 

- If so, we don't need to persist num of rows any more https://github.com/risingwavelabs/risingwave/pull/18983
- We will need a new code path in progress reporting. Cannot rely on the same code as MV backfill.
- This information is in state table (Can we just use state table for `SHOW JOBS`? And do not need to report progress any more 🤣 (still need report finish though.))

**Comment by xxchan** [comment]
But there's still a problem: we have per-partition offset. (i.e., in state table we have detailed per-split progress.) How to aggregate it for users?

**Comment by xxchan** [comment]
Some points:
- We don't always have the ability to calculate "Lag size" (or average backfilled offset / upstream offset)

	Note that, the "offset" may mean very different things in different connectors, unlike Kafka offset is as simple as array index.
	
	e.g., `Kinesis` offset is not consecutive. looks like `"SequenceNumber": "49639543389862895699365693304023657829928072185817071618"`. It can't be averaged, and we have no idea about the progress just from the sequence number.

	See also https://arc.net/l/quote/rdvyjwyz (#18299)

	Therefore, `num_backfilled_rows` might still be a valuable thing.
- We actually have a metric on Grafana: "Kafka Consumer Lag Size".
	
	Not sure whether we can get similar things for other connectors like Kinesis.
- Most of the time, backfill should be finished very quickly and not many users will care about observability of the backfill progress. (And we do have some advanced tricks for advanced users as mentioned above.) So I'm thinking it may not worth the effort to implement it perfectly. 🤪 Of course we can gradually improve it later.

**Comment by kwannoel** [comment]
For kinesis offset, I think as long as long as it is monotonically increasing, we can still use its average.

**Comment by xxchan** [comment]
@kwannoel Do you think we can merge it in current form and refine later?

**Review by kwannoel — APPROVED** [review]

**Pull request merged**